### PR TITLE
Deprecate public access to many mathtext internals.

### PR DESCRIPTION
**Pull request by anntzer**
For most of these classes, there isn't really a "reasonable" way to use
them via public API anyways.

- `Fonts` subclasses are looked up via the private `_font_type_mapping`
  and extending that would require adding new entries for
  `rcParams["mathtext.fontset"]`.
- `FontConstantsBase`\s are looked up via the private
  `_font_constant_mapping`.
- `Node` and its subclasses are clearly internal constructs of the
  mathtext parser (technically, they are visible to `MathtextBackend`
  subclasses when their `render` method is called, but one cannot
  register new `MathtextBackend` subclasses either as that happens via
  the private `_backend_mapping`).
- `Ship` is just an implementation detail for the `ship` callable, which
  can only be called on a `Node` or in a `MathtextBackend`, neither of
  which are really public per the above.
- `Parser` is set up and driven via the (still public) `MathTextParser`.
- `Error`, `SHRINK_FACTOR`, `GROW_FACTOR`, `NUM_SIZE_LEVELS` are helpers
  for the parser.
- `latex_to_*`, `stix_virtual_fonts`, `tex2uni` are helper lookup tables
  for the fonts.  (`tex2uni` stays indirectly accessible via the
  `get_unicode_index`, which basically performs lookup on it.)

Making these private should simplify the implementation of various
improvements to mathtext, e.g. better support for math fonts, as well as
code cleanups.

We can later decide how to properly expose these internals (e.g. #9165).

See also #16181.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: doc/api/next_api_changes/deprecations/18378-AL.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Deprecation of mathtext internals
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The following API elements previously exposed by the :mod:`.mathtext` module
+are considered to be implementation details of and public access to them is
```
Comment (timhoffm) [inline doc/api/next_api_changes/deprecations/18378-AL.rst:4]:
```suggestion
are considered to be implementation details and public access to them is
```

Comment (anntzer) [reply doc/api/next_api_changes/deprecations/18378-AL.rst:4]: yes

**Review by timhoffm — APPROVED** [review]
Conditional on CI passing.

File: lib/matplotlib/_mathtext.py [hunk 2]
```diff
@@ -0,0 +1,2967 @@
+"""
+Implementation details for :mod:`.mathtext`.
+"""
+
+from collections import namedtuple
+import functools
+import logging
+from io import StringIO
```
Comment (QuLogic) [inline lib/matplotlib/_mathtext.py:8]: Not sure why these changed order.

Comment (anntzer) [reply lib/matplotlib/_mathtext.py:8]: fixed, likely just a mistake

**Comment by anntzer** [comment]
I decided to also make `ship` private (now listed above).

File: doc/missing-references.json [hunk 3]
```diff
@@ -195,6 +200,70 @@
     "dviread.PsfontsMap": [
       "lib/matplotlib/dviread.py:docstring of matplotlib.dviread.DviFont:20"
     ],
+    "matplotlib._mathtext.Box": [
+      "doc/api/mathtext_api.rst:11",
```
Comment (QuLogic) [inline doc/missing-references.json:204]: It makes more sense to update these references, or delete them, rather than start accepting them as missing.

Comment (anntzer) [reply doc/missing-references.json:204]: The problem is that (AFAICT) at least they will always show up in the SVG inheritance diagram.  I got rid of the other occurrences, though.

File: doc/api/next_api_changes/deprecations/18378-AL.rst (hunk 1 above)
Comment (dopplershift) [reply doc/api/next_api_changes/deprecations/18378-AL.rst:4]: Looks like this change got lost?

File: lib/matplotlib/_mathtext.py (hunk 2 above)
Comment (QuLogic) [reply lib/matplotlib/_mathtext.py]: It's still flipped, so I don't know why GitHub says it's outdated.

File: doc/missing-references.json [hunk 4]
```diff
@@ -616,19 +724,25 @@
...
     ],
     "Figure.stale_callback": [
-      "doc/users/interactive_guide.rst:336"
+      "doc/users/interactive_guide.rst:333"
     ],
     "FigureCanvas": [
       "lib/matplotlib/backend_tools.py:docstring of matplotlib.backend_tools.ToolBase:25"
     ],
+    "Fonts": [
+      "doc/api/next_api_changes/deprecations/18004-AL.rst:3"
```
Comment (QuLogic) [inline doc/missing-references.json:733]: This can be fixed.

File: lib/matplotlib/_mathtext.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/_mathtext.py]: let's do it again :/

File: doc/missing-references.json (hunk 4 above)
Comment (anntzer) [reply doc/missing-references.json:733]: yup

Comment (QuLogic) [reply doc/missing-references.json:733]: So this entry can be removed now?

Comment (anntzer) [reply doc/missing-references.json]: third try's the charm?

**Review by QuLogic — APPROVED** [review]

File: doc/missing-references.json (hunk 4 above)
Comment (QuLogic) [reply doc/missing-references.json]: Indeed.

**Pull request merged**